lib/r10k/module_loader/puppetfile.rb

```diff
@@ -0,0 +1,190 @@
+          args = { version: args }
+        end
+
+        args[:overrides] = @overrides
+
+        if install_path = args.delete(:install_path)
+          install_path = resolve_install_path(install_path)
+          validate_install_path!(install_path, name)
+        else
+          install_path = @moduledir
+        end
+
+        if @default_branch_override != nil
+          args[:default_branch_override] = @default_branch_override
+        end
+
+        mod = R10K::Module.new(name, install_path, args, @environment)
+        mod.origin = :puppetfile
+
+        # Do not load modules if they would conflict with the attached
+        # environment
+        if @environment && @environment.module_conflicts?(mod)
+          mod = nil
+          return @modules
+        end
+
+        @modules << mod
+      end
+
+     private
+
+      # @param [Array<R10K::Module>] modules
+      def validate_no_duplicate_names!(modules)
+        dupes = modules
+                .group_by { |mod| mod.name }
+                .select { |_, v| v.size > 1 }
+                .map(&:first)
+        unless dupes.empty?
+          msg = _('Puppetfiles cannot contain duplicate module names.')
+          msg += ' '
+          msg += _("Remove the duplicates of the following modules: %{dupes}" % { dupes: dupes.join(' ') })
+          raise R10K::Error.new(msg)
+        end
+      end
+
+      def resolve_install_path(path)
+        pn = Pathname.new(path)
+
+        unless pn.absolute?
+          pn = Pathname.new(File.join(@basedir, path))
+        end
+
+        # .cleanpath is as good as we can do without touching the filesystem.
+        # The .realpath methods will also choke if some of the intermediate
+        # paths are missing, even though we will create them later as needed.
+        pn.cleanpath.to_s
+      end
+
+      def validate_install_path!(path, modname)
+        unless /^#{Regexp.escape(real_basedir)}.*/ =~ path
+          raise R10K::Error.new("Puppetfile cannot manage content '#{modname}' outside of containing environment: #{path} is not within #{real_basedir}")
+        end
+
+        true
+      end
+
+      def determine_managed_directories(managed_content)
+        managed_content.keys.reject { |dir| dir == real_basedir }
+      end
+
+      # Returns an array of the full paths to all the content being managed.
+      # @return [Array<String>]
+      def determine_desired_contents(managed_content)
+        managed_content.flat_map do |install_path, mods|
+          mods.collect { |mod| File.join(install_path, mod.name) }
+        end
+      end
+
+      def determine_purge_exclusions(managed_dirs)
+        if environment && environment.respond_to?(:desired_contents)
+          managed_dirs + environment.desired_contents
+        else
+          managed_dirs
+        end
+      end
+
+      def real_basedir
+        Pathname.new(@basedir).cleanpath.to_s
+      end
+
+      # For testing purposes only
+      def puppetfile_content(path)
```

Comment:
yeah, I'm just not really understanding why that's even necessary. This isn't acting on anything within this class.. 

In this example, you're acting on `puppetfile`, but this is the `loader` class.. Should this method go in the `R10K::Puppetfile` class? Or should this not take a param and instead call `File.read(@puppetfile)`?

Reply:
So the fact that it takes a param instead of acting on an instance variable is the confusing part? It's six of one, half dozen of the other to me. Though I typically like methods parameterized as much as possible, which is why it has a param.

Here is an example of how this is used in a test: https://github.com/puppetlabs/r10k/pull/1167/files#diff-1bc365ddca4e6e4a7ba95ff4d655c036faa16ec1025e35bc559476265a5719aaR292

Reply:
No, that's not what bothering/confusing me.

I was mostly objecting to the premise of adding a class method just for testing purposes, especially since I wasn't really understanding how calling it in the tests was actually doing anything, but now that I look again, I see that this method actually _is_ being used within the `load` method, not just in tests, in which case I think you should remove this comment, since it clearly confused me a lot lol.

Reply:
I agree this makes things better for testing, it gets messy stubbing things like `File.read` because once you do it once in a test, you have to do it for all instances of it being called, which often means stubbing a lot more than is strictly necessary. So far as I can tell, it _is_ called up above, where it's being passed the `puppetfile` instance var. The little abstraction over `File.read` in that context is kinda nice actually, IMO.

Reply:
yeah, I'm totally fine with this method now that I see how it's being used. The "for testing purposes only" comment totally threw me off.